### ISSUE TITLE
WU download transfer failure, Retry fails, results in downloading a new WU

### ISSUE THREAD

**Issue by Hou5e** [FoldingAtHome/fah-client-bastet#70]
Linux Mint 21.1 (U22.04), FAH Client v8.1.6:
Single download WU failure, results in 2 log error messages. The WU that fails to download is abandoned or dumped (WU17) and requests a new WU (WU18). The second error message (Can't dereference NULL pointer!) sounds like the code could handle the error better.

(I removed the other GPU's messages)
00:52:42:I1::Added new work unit: cpus:1 gpus:gpu:02:00:00
00:52:42:I1::WU17:Requesting WU assignment
00:52:42:I1:OUT35:> POST https://assign6.foldingathome.org/api/assign HTTP/1.1
00:52:42:I3:Connecting to assign6.foldingathome.org:443
00:52:43:I1:OUT35:< assign6.foldingathome.org:443 HTTP/1.1 200 HTTP_OK
00:52:43:I1::WU17:Received WU assignment F1JFxb2pIwTSOHwhVbigmp2NC4de0eA_bzjnA0HjLTc
00:52:43:I1::WU17:Downloading WU
00:52:43:I1:OUT37:> POST https://ds03.scs.illinois.edu/api/assign HTTP/1.1
00:52:43:I3:Connecting to ds03.scs.illinois.edu:443
**_00:53:32:E ::WU17:Failed response: EOF_**
**_00:53:32:I1::WU17:Retrying in 2.00 secs_**
00:53:34:I1::WU17:Downloading WU
00:53:34:I1:OUT38:> POST https://ds03.scs.illinois.edu/api/assign HTTP/1.1
00:53:34:I3:Connecting to ds03.scs.illinois.edu:443
**_00:53:35:I1::WU17:DOWNLOAD 100%_**
00:53:35:I1:OUT38:< ds03.scs.illinois.edu:443 HTTP/1.1 500 HTTP_INTERNAL_SERVER_ERROR
**_00:53:35:E ::WU17:HTTP_INTERNAL_SERVER_ERROR: {"error":{"message":"Can't dereference NULL pointer!"}}_**
00:53:37:I1::Added new work unit: cpus:1 gpus:gpu:02:00:00
**_00:53:37:I1::WU18:Requesting WU assignment_**
00:53:37:I1:OUT39:> POST https://assign1.foldingathome.org/api/assign HTTP/1.1
00:53:37:I3:Connecting to assign1.foldingathome.org:443
00:53:38:I1:OUT39:< assign1.foldingathome.org:443 HTTP/1.1 200 HTTP_OK
00:53:38:I1::WU18:Received WU assignment AFg2tbjoALc5TVP2MdHRY3s-LrqPRI8tcvje0dV_BBc
00:53:38:I1::WU18:Downloading WU
00:53:38:I1:OUT40:> POST https://ds03.scs.illinois.edu/api/assign HTTP/1.1
00:53:38:I3:Connecting to ds03.scs.illinois.edu:443
00:53:41:I1::WU18:DOWNLOAD 94%
00:53:41:I1::WU18:DOWNLOAD 100%
00:53:41:I1:OUT40:< ds03.scs.illinois.edu:443 HTTP/1.1 200 HTTP_OK

**Comment by kbernhagen** [comment]
Is it correct to say the unit was abandoned, not dumped?

**Comment by Hou5e** [comment]
The work was not in the Work folder, so it was probably dumped. There was no 'Dumping WU' message in the log though.

**Comment by Hou5e** [comment]
When a WU is dumped, you get a file in the 'credits' folder (grep -i 'dumped' ./*.json) with the hash name of the WU assignment. There was no credits file: 'F1JFxb2pIwTSOHwhVbigmp2NC4de0eA_bzjnA0HjLTc.json' with a 'dumped' message.

So, I would say the WU was abandoned, and the retry to download the same original assigned WU might have been hosed from not clearing the first error completely before retying to download (just a guess without looking at the code).

**Comment by Hou5e** [comment]
[different Linux Mint 21.1 PC] Different download error, and the same retry failure (most likely from the previous error not being completely reset before the retry)

01:19:49:I1::Added new work unit: cpus:1 gpus:gpu:02:00:00
01:19:49:I1::WU28:Requesting WU assignment
01:19:49:I1:OUT53:> POST https://assign6.foldingathome.org/api/assign HTTP/1.1
01:19:49:I3:Connecting to assign6.foldingathome.org:443
01:19:50:I1:OUT53:< assign6.foldingathome.org:443 HTTP/1.1 200 HTTP_OK
01:19:50:I1::WU28:Received WU assignment Wgj1wXSe1pm5klu1FGHLI8VvLCqxIpvUI1Aaav-iVNw
01:19:50:I1::WU28:Downloading WU
01:19:50:I1:OUT55:> POST https://ds03.scs.illinois.edu/api/assign HTTP/1.1
01:19:50:I3:Connecting to ds03.scs.illinois.edu:443
01:19:51:I1::WU28:DOWNLOAD 100%
01:19:51:I1:OUT55:< ds03.scs.illinois.edu:443 HTTP/1.1 503 HTTP_SERVICE_UNAVAILABLE
**_01:19:51:E ::WU28:HTTP_SERVICE_UNAVAILABLE: {"error":{"message":"Please wait","code":503}}_**
**_01:19:51:I1::WU28:Retrying in 2.00 secs_**
01:19:53:I1::WU28:Downloading WU
01:19:53:I1:OUT56:> POST https://ds03.scs.illinois.edu/api/assign HTTP/1.1
01:19:53:I3:Connecting to ds03.scs.illinois.edu:443
01:19:54:I1::WU28:DOWNLOAD 100%
01:19:54:I1:OUT56:< ds03.scs.illinois.edu:443 HTTP/1.1 500 HTTP_INTERNAL_SERVER_ERROR
**_01:19:54:E ::WU28:HTTP_INTERNAL_SERVER_ERROR: {"error":{"message":"Can't dereference NULL pointer!"}}_**
01:19:56:I1::Added new work unit: cpus:1 gpus:gpu:02:00:00
**_01:19:56:I1::WU29:Requesting WU assignment_**
01:19:56:I1:OUT57:> POST https://assign1.foldingathome.org/api/assign HTTP/1.1
01:19:56:I3:Connecting to assign1.foldingathome.org:443
01:19:57:I1:OUT57:< assign1.foldingathome.org:443 HTTP/1.1 200 HTTP_OK
01:19:57:I1::WU29:Received WU assignment ZGwAofYxAOw4hEHDjvcYQTUH89PtUkQ8uh65Mk8Vmjo
01:19:57:I1::WU29:Downloading WU
01:19:57:I1:OUT58:> POST https://ds03.scs.illinois.edu/api/assign HTTP/1.1
01:19:57:I3:Connecting to ds03.scs.illinois.edu:443
01:20:00:I1::WU29:DOWNLOAD 100%
01:20:00:I1:OUT58:< ds03.scs.illinois.edu:443 HTTP/1.1 200 HTTP_OK
01:20:00:I1::WU29:Received WU

**Comment by jcoffland** [comment]
If the client is unable to download a WU it is correct for it to abandon the assignment and request a new one from the AS.  If there was an error like ``HTTP_INTERNAL_SERVER_ERROR: {"error":{"message":"Can't dereference NULL pointer!"}}`` on the WS then it will not have completed the assignment.  I think this is the correct behavior of the client when the WS fails to assign.

I looked at the WS logs and the ``Can't dereference NULL pointer!`` error was triggered by the client retrying the assignment after ``HTTP_SERVICE_UNAVAILABLE``.  I believe this problem has already been fixed in the latest WS code.  Many servers still need to be upgraded though.
